File: android/app/src/main/java/io/github/magnusencoded/setlist2spotify/data/spotify/SpotifyModels.kt
package io.github.magnusencoded.setlist2spotify.data.spotify

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class TokenResponse(
    @SerialName("access_token") val accessToken: String,
    @SerialName("expires_in") val expiresIn: Long = 3600,
    @SerialName("refresh_token") val refreshToken: String? = null,
    val scope: String = "",
)

@Serializable
data class SpotifyUser(
    val id: String,
    @SerialName("display_name") val displayName: String? = null,
)

@Serializable
data class TrackSearchResponse(
    val tracks: TrackPage? = null,
)

@Serializable
data class TrackPage(
    val items: List<SpotifyTrack> = emptyList(),
)

@Serializable
data class SpotifyTrack(
    val id: String,
    val name: String,
    val uri: String,
    val artists: List<SpotifyArtist> = emptyList(),
    val album: SpotifyAlbum? = null,
    @SerialName("duration_ms") val durationMs: Long = 0,
) {
    fun artistNames(): String = artists.joinToString(", ") { it.name }
}

@Serializable
data class SpotifyArtist(
    val name: String = "",
)

@Serializable
data class SpotifyAlbum(
    val name: String? = null,
)

@Serializable
data class PlaylistResponse(
    val id: String,
    @SerialName("external_urls") val externalUrls: Map<String, String> = emptyMap(),
)

File: ios/StationToStation/Data/Spotify/SpotifyModels.swift
import Foundation

struct TokenResponse: Decodable {
    @DefaultCodable<EmptyString> var accessToken = ""
    @DefaultCodable<ExpiresDefault> var expiresIn: Double = 3600
    var refreshToken: String?
    @DefaultCodable<EmptyString> var scope = ""

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case expiresIn = "expires_in"
        case refreshToken = "refresh_token"
        case scope
    }
}

struct SpotifyUser: Decodable {
    @DefaultCodable<EmptyString> var id = ""
    var displayName: String?
    var product: String?

    enum CodingKeys: String, CodingKey {
        case id
        case displayName = "display_name"
        case product
    }
}

struct TrackSearchResponse: Decodable {
    var tracks: TrackPage?
}

struct TrackPage: Decodable {
    @DefaultCodable<EmptyArray<SpotifyTrack>> var items: [SpotifyTrack] = []
}

struct SpotifyTrack: Decodable, Identifiable {
    @DefaultCodable<EmptyString> var id = ""
    @DefaultCodable<EmptyString> var name = ""
    @DefaultCodable<EmptyString> var uri = ""
    @DefaultCodable<EmptyArray<SpotifyArtist>> var artists: [SpotifyArtist] = []
    var album: SpotifyAlbum?
    @DefaultCodable<LongZero> var durationMs: Int64 = 0

    enum CodingKeys: String, CodingKey {
        case id, name, uri, artists, album
        case durationMs = "duration_ms"
    }

    func artistNames() -> String { artists.map(\.name).joined(separator: ", ") }
}

struct SpotifyArtist: Decodable {
    @DefaultCodable<EmptyString> var name = ""
}

struct SpotifyAlbum: Decodable {
    var name: String?
}

struct PlaylistResponse: Decodable {
    @DefaultCodable<EmptyString> var id = ""
    @DefaultCodable<EmptyStringMap> var externalUrls: [String: String] = [:]

    enum CodingKeys: String, CodingKey {
        case id
        case externalUrls = "external_urls"
    }
}

/// A playlist as returned by GET /playlists/{id}; carries who made it and its description.
struct SimplePlaylist: Decodable {
    @DefaultCodable<EmptyString> var id = ""
    @DefaultCodable<EmptyString> var name = ""
    var description: String?
    var owner: SpotifyUser?
}
